File: new/74_search_a_2d_matrix_binary_search_one_pass.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        # idea: binary search (one go)
        # convert the 2d array to 1d array
        # iteratively convert to and fro from middle index of the 1D array
        # and row and col of 2d array to compare target and middle value
        # and move left and right pointer accordingly.
        ROW = len(matrix)
        COL = len(matrix[0])

        left = 0
        right = (ROW * COL) - 1  # rows * cols gives us the last index in 1d

        while left <= right:
            middle = left + (right - left) // 2
            # using this middle recalculate current value from 2d array to compare with target
            middle_row, middle_col = middle // COL, middle % COL
            # index // COL gives us which row (how many full rows can fit before current row)
            # idex % COL gives us which column of the given row
            current_val = matrix[middle_row][middle_col]
            if target < current_val:
                right = middle - 1
            elif target > current_val:
                left = middle + 1
            else:
                return True

        return False
        # time: O(log(m * n))
        # space: O(1)

File: new/cpp/74_search_a_2d_matrix_binary_search_one_pass.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        // idea: binary search (one pass)
        // flatten the 2d array into 1d array (treat the 2d as 1d by manipulating indices)
        // convert to and fro indices, finding middle element of 1d array, and converting that
        // middle index to the appropriate index of the 2d array and find the value of that index
        // and using that to compare that to the target
        // do this iteratively until we find the answer or not. 
        int ROWS = matrix.size();
        int COLS = matrix[0].size();

        int left = 0;
        int right = (ROWS * COLS) - 1;

        while (left <= right) {
            int middle = left + (right - left) / 2; 
            int middle_row = middle / COLS;
            int middle_col = middle % COLS;
            int current_val = matrix[middle_row][middle_col];
            if (target > current_val) {
                left = middle + 1;
            } else if (target < current_val) {
                right = middle - 1;
            } else {
                return true;
            }
        }

        return false;
    }
    // time: O(log(m * n))
    // space: O(1)
};
